### PR TITLE
Versioned thresholds

### LINKED ISSUE #228: Allow thresholds checking based on version
Somewhat related to
" Allow whitelisting of errors/warnings based on version #106"
as we have a similar problem there.

### PR DESCRIPTION
**Pull request by Karm**
fixes #228

# Thresholds properties

We need to switch on and off certain tests depending on native-image versions used,
on host vs. container and most importantly, based on Quarkus version.

We use method annotations for that, e.g. `@IfQuarkusVersion(min = "3.6.0")`,
or `@IfMandrelVersion(min = "23.0.0", inContainer = true)`.

We also use properties in text files to validate whether particular thresholds were crossed, e.g. whether the app used more RAM than expected or whether the measured time delta between JVM HotSpot mode and native-image mode was bigger than expected.

e.g.

```
linux.jvm.time.to.finish.threshold.ms=6924
linux.native.time.to.finish.threshold.ms=14525
```

The current `.conf` format enhances `.properties` format with the power of using the 
annotation strings, see:

```
# Comments and empty lines are ignored
linux.jvm.time.to.finish.threshold.ms=6000
linux.native.time.to.finish.threshold.ms=14000
linux.executable.size.threshold.kB=79000

@IfQuarkusVersion(min ="2.7.0", max="3.0.0")
linux.jvm.time.to.finish.threshold.ms=5000
linux.native.time.to.finish.threshold.ms=13000
linux.executable.size.threshold.kB=75000

@IfQuarkusVersion(min ="3.5.0", max="3.5.999")
@IfMandrelVersion(min = "23.1.2", minJDK = "21.0.1" )
linux.jvm.time.to.finish.threshold.ms=6924
linux.native.time.to.finish.threshold.ms=14525
linux.executable.size.threshold.kB=79000

@IfQuarkusVersion(min ="3.6.0")
linux.jvm.time.to.finish.threshold.ms=6924
linux.native.time.to.finish.threshold.ms=14525
linux.executable.size.threshold.kB=79000

@IfMandrelVersion(min = "24", minJDK = "21.0.1" )
linux.executable.size.threshold.kB=90000
```

Properties are being added to a map top to bottom, overwriting their previous values
unless an `@If` constraint fails. If a condition fails, the following properties are
not added to the map until the next `@If` constraint is met.

If two `@If` constraints follow immediately one after the other, they both MUST be true
to process the following properties.

Take a look at [ThresholdsTest.java](./ThresholdsTest.java) and its `threshold-*.conf` test [files](../../../../../../../../test/resources/) for a comprehensive overview.

The parsing logic is compatible with plain `.properties` files as we have been using before,
i.e. any key-value pair where the value is interpreted as the long type.

### CODE REVIEW

**Comment by Karm** [comment]
eh, gonna fix WInodws

**Comment by Karm** [comment]
The problem is that my code is now forcing inContainer version check even on a system where container might not be available, e.g. Windows.   I think I'd just drop  the ability to specify threshold for container apps specifically. We haven't had container specific thresholds anyway to date I'd say.

File: testsuite/src/it/java/org/graalvm/tests/integration/utils/Apps.java [hunk 1]
```diff
@@ -178,34 +181,31 @@ public enum Apps {
...
+        if (Files.exists(tcFile)) {
+            final String appDirNormalized = dir.toUpperCase().replace(File.separator, "_").replace('-', '_') + "_";
+            try {
+                final Map<String, Long> props = parseProperties(tcFile);
+                for (String pn : props.keySet()) {
+                    final String normPn = pn.toUpperCase().replace('.', '_');
+                    final String env = System.getenv().get(appDirNormalized + normPn);
                     if (StringUtils.isNotBlank(env)) {
-                        props.replace(pn, env);
+                        props.replace(pn, Long.parseLong(env));
```
Comment (zakkak) [inline testsuite/src/it/java/org/graalvm/tests/integration/utils/Apps.java:195]: Do we really want to keep the ability to override thresholds? What's the use case?

File: testsuite/src/it/java/org/graalvm/tests/integration/utils/versions/MandrelVersionCondition.java [hunk 2]
```diff
@@ -42,6 +43,8 @@ public class MandrelVersionCondition implements ExecutionCondition {
     private static final ConditionEvaluationResult ENABLED_BY_DEFAULT =
             enabled("@IfMandrelVersion is not present");
 
+    private static final Pattern jdkPattern = Pattern.compile("(?<jfeature>[0-9]+)(\\.(?<jinterim>[0-9]*)\\.(?<jupdate>[0-9]*))?");
```
Comment (zakkak) [inline testsuite/src/it/java/org/graalvm/tests/integration/utils/versions/MandrelVersionCondition.java:46]: Nitpick: consider renaming field:

```suggestion
    private static final Pattern JDK_PATTERN = Pattern.compile("(?<jfeature>[0-9]+)(\\.(?<jinterim>[0-9]*)\\.(?<jupdate>[0-9]*))?");
```

**Review by zakkak — APPROVED** [review]
LGTM, please have a look at my comments for some suggestions on potential improvements.

File: testsuite/src/it/java/org/graalvm/tests/integration/utils/Apps.java (hunk 1 above)
Comment (Karm) [reply testsuite/src/it/java/org/graalvm/tests/integration/utils/Apps.java:195]: I am not sure people even knew it was there.
I used to use it back in the day in internal Jenkins jobs where I overwrote certain threshold according to the particular executor (stronger vs. weaker vm) as a JOB env parameter.

I don't think it is being actively used anywhere now...

Comment (Karm) [reply testsuite/src/it/java/org/graalvm/tests/integration/utils/Apps.java:195]: Ha, it's been documented all along though: 

https://github.com/Karm/mandrel-integration-tests#thresholds

File: testsuite/src/it/java/org/graalvm/tests/integration/utils/thresholds/Thresholds.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+                        // We do not re-evaluate the IfMandrel line.
+                        i++;
+                    } else {
+                        useProp = ifQuarkus;
+                    }
+                } else {
+                    useProp = ifQuarkus;
+                }
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (Karm) [reply testsuite/src/it/java/org/graalvm/tests/integration/utils/thresholds/Thresholds.java:132]: Thanks!

**Comment by Karm** [comment]
@zakkak  notes addressed, merging

**Pull request merged**